src/main/java/rx/observables/StringObservable.java

```diff
@@ -114,28 +117,106 @@ public void call(S resource) {
+                    // already started with fast path
+                    return;
+                else if (n == Long.MAX_VALUE) {
+                    // fast path
+                    requestAll();
+                } else 
+                    requestSome(n);
+            } catch (RuntimeException e) {
+                subscriber.onError(e);
+            } catch (IOException e) {
+                subscriber.onError(e);
+            }
+        }
+
+        private void requestAll() {
+            requested.set(Long.MAX_VALUE);
+            byte[] buffer = new byte[size];
+            try {
+                if (subscriber.isUnsubscribed())
+                    return;
+                int n = is.read(buffer);
+                while (n != -1 && !subscriber.isUnsubscribed()) {
+                    subscriber.onNext(Arrays.copyOf(buffer, n));
+                    if (!subscriber.isUnsubscribed())
+                        n = is.read(buffer);
+                }
+            } catch (IOException e) {
+                subscriber.onError(e);
+            }
+            if (subscriber.isUnsubscribed())
+                return;
+            subscriber.onCompleted();
+        }
+        
+        
+
+        private void requestSome(long n) throws IOException {
+            // back pressure path
+            // this algorithm copied roughly from
+            // rxjava/OnSubscribeFromIterable.java
+
+            long previousCount = requested.getAndAdd(n);
+            if (previousCount == 0) {
                 byte[] buffer = new byte[size];
-                try {
-                    if (o.isUnsubscribed())
+                while (true) {
+                    long r = requested.get();
+                    long numToEmit = r;
+
+                    //emit numToEmit
+                    
+                    if (subscriber.isUnsubscribed())
                         return;
-                    int n = i.read(buffer);
-                    while (n != -1 && !o.isUnsubscribed()) {
-                        o.onNext(Arrays.copyOf(buffer, n));
-                        if (!o.isUnsubscribed())
-                            n = i.read(buffer);
+                    int numRead;
+                    if (numToEmit>0)
+                        numRead = is.read(buffer);
+                    else 
+                        numRead = 0;
+                    while (numRead != -1 && !subscriber.isUnsubscribed() && numToEmit>0) {
+                        subscriber.onNext(Arrays.copyOf(buffer, numRead));
+                        numToEmit--;
+                        if (numToEmit >0 && !subscriber.isUnsubscribed())
+                            numRead = is.read(buffer);
                     }
-                } catch (IOException e) {
-                    o.onError(e);
+                    
+                    //check if we have finished
+                    if (numRead == -1 && !subscriber.isUnsubscribed()) 
+                        subscriber.onCompleted();
```

Comment:
A ```return``` is missing here; without it, the while loop would spin until the downstream actually unsubscribes.